### PR TITLE
tetragon: Allow to use data events for char_buf data

### PR DESCRIPTION
**Pull request by olsajiri**
Adding support to specify 'maxData: true' flag for char_buf argument
type to retrieve maximum data.

Using data events (same as for exec sensor) to retrieve and send
data to the user space from bpf program.

It's available for large kernels. Kernels < 5.3 won't load the 
generic kprobe with data events due to verifier size limit.

It's possible to use the maxData flag like:
```
  - call: "__x64_sys_write"
    return: false
    syscall: true
    args:
    - index: 0
      type: "int"
    - index: 1
      type: "char_buf"
      sizeArgIndex: 3
      maxData: true
    - index: 2
      type: "size_t"
```
The data events are used only if the data won't fit into standard
event size.

There's no return kprobe support at the moment. I'm adding this
for terminal sniff, where it's not needed.

The maximum size for char_buf argument with maxData flag is 327360 bytes.

### CODE REVIEW

**Comment by olsajiri** [comment]
static check fails on:
`CHECK:BIT_MACRO: Prefer using the BIT macro`

can be ignored

**Review by kkourt — CHANGES_REQUESTED** [review]
Thanks!

Can we add a warning for the cases where the configuration will be ignored from bpf side?

File: pkg/sensors/tracing/generickprobe.go [hunk 1]
```diff
@@ -187,6 +190,9 @@ func getMetaValue(arg *v1alpha1.KProbeArg) (int, error) {
 	if arg.ReturnCopy {
 		meta = meta | argReturnCopyBit
 	}
+	if arg.Max {
+		meta = meta | argMaxBit
```
[the remark this answers is not in the document]
Comment (olsajiri) [reply pkg/sensors/tracing/generickprobe.go:188]: makes sense, I just pushed the change

**Review by mtardy — COMMENTED** [review]
Do you think you could document this new feature in the TracingPolicy reference, in the argument paragraph, where we already talk about `char_buf` and `char_iovec`: https://tetragon.cilium.io/docs/reference/tracing-policy/#arguments ? Like how this thing interacts with `returnCopy` maybe as well, what's the difference.

File: pkg/sensors/tracing/kprobe_test.go [hunk 2]
```diff
@@ -2599,3 +2603,210 @@ spec:
...
+	_, fd2, fdString := createTestFile(t)
+	myPid := observer.GetMyPid()
+	pidStr := strconv.Itoa(int(myPid))
+	writeHook := `
+apiVersion: cilium.io/v1alpha1
+metadata:
+  name: "sys-write"
+spec:
+  kprobes:
+  - call: "__x64_sys_write"
```
[the remark this answers is not in the document]
Comment (olsajiri) [reply pkg/sensors/tracing/kprobe_test.go:2710]: ok

File: pkg/sensors/tracing/kprobe_test.go [hunk 3]
```diff
@@ -2599,3 +2603,210 @@ spec:
...
```
[the remark this answers is not in the document]
Comment (olsajiri) [reply pkg/sensors/tracing/kprobe_test.go:2768]: ok

**Comment by olsajiri** [comment]
I rebased new version on top of https://github.com/cilium/tetragon/pull/885

**Comment by kkourt** [comment]
> I rebased new version on top of #885

Thanks! It seems that it needs one more rebase now that #885 was merged :(

**Review by kkourt — APPROVED** [review]

**Review by kkourt — APPROVED** [review]

**Pull request merged**